rfcs/1440_generic_planner_node.md

```diff
@@ -0,0 +1,482 @@
+message TagTemplate {
+  oneof tag {
+    GrpcTag grpc_tag = 1;
+    GrpcTagTemplate grpc_tag_template = 1;
+    WebAssemblyModuleTag web_assembly_module_tag = 2;
+    WebAssemblyModuleSignatureTag web_assembly_module_signature_tag = 3;
+    TlsEndpointTag tls_endpoint_tag = 4;
+  }
+}
+
+```
+
+As an example, the following can be used to define a label template matching a
+user-specific gRPC confidentiality tag and extracting the byte-array content of
+the tag as a parameter named user_id:
+
+```toml
+[[label_template.confidentiality_tags]]
+
+[label_template.confidentiality_tags.tag.grpc_tag_template]
+user_id_parameter_name = "user_id"
+
+```
+
+Note: I am unsure about the exact TOML representation for this.
+
+Question: In general TOML is not great for deeply nested object hierarchies, so
+we should perhaps consider a different serialisation format. Prost does not
+support text-serialised protos. JSON is an option, but not that great for human
+readable config. YAML can handle the complex object hierarchies, but the
+specification is very complex and parsing YAML is considered risky due to
+potential of injection vulnerabilities that is common when trying to support
+dynamic local object creation as defined by the specification. The application
+configuration is in TOML format, so having to deal with multiple serialisation
+formats when creating an application is not a great developer experience. Is
+there a suitable format that will work for everything?
+
+### Invocation Routing
+
+The planner node will be responsible for creating the required pseudo nodes as
+well. In the specific cases of HTTP- and gRPC pseudo nodes each pseudo node will
+send all invocations over a single dedicated invocation channel. The planner
+node will, by default, have the handle to receive these invocations and
+therefore be responsible for routing them to the right Wasm nodes based on the
+associated labels (specifically the label associated with the request receiver
+channel should match the label of the target Wasm node). Since it would be
+possible for multiple Wasm nodes to share the same label, the planner node would
+need an indication of which Wasm node should receive invocations with a specific
+label on the request receiver channel.
+
+If more advanced routing is required, the planner node could forward the
+invocation receiver channel to another node that can do the routing. The problem
+in this case is that the new node would have to be responsible for creating any
+per-label nodes as well if these nodes must be dynamically created, seeing that
+the planner node will no longer be able to see incoming invocations.
+
+Questions:
+
+- Would there be a requirement for multiple gRPC server pseudo nodes running on
+  different ports to each have a different configuration for Invocation routing?
+- Should we use special handling of channel types to determine the targets for
+  routing, or should we mark specific nodes as targeted invocation receivers?
```

Comment:
This seems too specific, I think we just need a mechanism to define a graph, and instantiate it by providing concrete labels to bind to label parameters. Routing logic, what to do about invocations, how to create the initial gRPC and HTTP nodes etc. may still be driven by regular Wasm code.

Reply:
The point is to restrict node creation so that we can reason about what the node graph will look like without having to analyse arbitrary rust code to try and figure out all possible node graphs that might ever be created.

Reply:
Unless there is a concrete definition of what is "safe" and what not, this seems to just add more complexity in the hope that bad things do not happen. But I think it would be good to first define the category of things that we expect this system to catch, and then show how it actually catches them, and make sure we only add the minimum amount of extra logic to prevent things that we know are undesirable.